File: dashboard/src/lib/table.ts
import { useSyncExternalStore } from "react";
import type { Database as SqlJsDatabase } from "sql.js";
import type { TableSchema } from "./types";

type Listener = () => void;

/** djb2 hash over a string → unsigned 32-bit integer. */
function djb2(str: string): number {
  let hash = 5381;
  for (let i = 0; i < str.length; i++) {
    hash = ((hash << 5) + hash + str.charCodeAt(i)) & 0xffffffff;
  }
  return hash >>> 0;
}

export class Table {
  readonly schema: TableSchema;
  readonly pkCol: string;
  private _db: SqlJsDatabase;
  private _version = 0;
  private _hash = 0;
  private _listeners: Set<Listener> = new Set();
  private _colNames: string[];

  constructor(schema: TableSchema, db: SqlJsDatabase) {
    const pkCols = schema.columns.filter((c) => c.primaryKey);
    if (pkCols.length !== 1) {
      throw new Error(
        `Table "${schema.name}" must have exactly one primaryKey column, got ${pkCols.length}`
      );
    }
    this.schema = schema;
    this.pkCol = pkCols[0].name;
    this._db = db;
    this._colNames = schema.columns.map((c) => c.name);
  }

  /** Generate and execute CREATE TABLE + CREATE INDEX statements. */
  create(): void {
    const colDefs = this.schema.columns.map((c) => {
      const parts = [c.name, c.type];
      if (c.primaryKey) parts.push("PRIMARY KEY");
      if (c.autoincrement) parts.push("AUTOINCREMENT");
      if (c.notNull) parts.push("NOT NULL");
      if (c.default !== undefined) parts.push(`DEFAULT (${c.default})`);
      return parts.join(" ");
    });

    this._db.run(
      `CREATE TABLE IF NOT EXISTS ${this.schema.name} (\n  ${colDefs.join(",\n  ")}\n);`
    );

    for (const idx of this.schema.indexes ?? []) {
      this._db.run(
        `CREATE INDEX IF NOT EXISTS ${idx.name} ON ${this.schema.name}(${idx.columns.join(", ")});`
      );
    }
  }

  /** Upsert a row (INSERT OR REPLACE). Returns lastInsertRowid. */
  upsert(row: Record<string, unknown>): number {
    const pkVal = row[this.pkCol];
    if (pkVal === undefined || pkVal === null) {
      throw new Error(`Row must include primary key column "${this.pkCol}"`);
    }
    // XOR out old row hash if it exists
    const existing = this._getByPk(pkVal);
    if (existing) {
      this._hash ^= this._hashRow(existing);
    }

    const cols = Object.keys(row);
    const placeholders = cols.map(() => "?").join(", ");
    const values = cols.map((c) => row[c]);
    this._db.run(
      `INSERT OR REPLACE INTO ${this.schema.name} (${cols.join(", ")}) VALUES (${placeholders})`,
      values as (string | number | null | Uint8Array)[]
    );

    // XOR in new row hash — read back the full row to get defaults
    const inserted = this._getByPk(pkVal);
    if (inserted) {
      this._hash ^= this._hashRow(inserted);
    }

    const result = this._db.exec("SELECT last_insert_rowid() as id");
    this._bump();
    return result.length > 0 ? (result[0].values[0][0] as number) : 0;
  }

  /** Batch upsert multiple rows. Single version bump + hash update. */
  upsertMany(rows: Record<string, unknown>[]): void {
    if (rows.length === 0) return;

    for (const row of rows) {
      const pkVal = row[this.pkCol];
      if (pkVal === undefined || pkVal === null) {
        throw new Error(`Row must include primary key column "${this.pkCol}"`);
      }
      const existing = this._getByPk(pkVal);
      if (existing) {
        this._hash ^= this._hashRow(existing);
      }

      const cols = Object.keys(row);
      const placeholders = cols.map(() => "?").join(", ");
      const values = cols.map((c) => row[c]);
      this._db.run(
        `INSERT OR REPLACE INTO ${this.schema.name} (${cols.join(", ")}) VALUES (${placeholders})`,
        values as (string | number | null | Uint8Array)[]
      );

      const inserted = this._getByPk(pkVal);
      if (inserted) {
        this._hash ^= this._hashRow(inserted);
      }
    }

    this._bump();
  }

  /** Select rows with optional WHERE, params, orderBy, columns. */
  select(opts?: {
    where?: string;
    params?: unknown[];
    orderBy?: string;
    columns?: string;
  }): Record<string, unknown>[] {
    let sql = `SELECT ${opts?.columns ?? "*"} FROM ${this.schema.name}`;
    if (opts?.where) sql += ` WHERE ${opts.where}`;
    if (opts?.orderBy) sql += ` ORDER BY ${opts.orderBy}`;

    const results = this._db.exec(sql, opts?.params as (string | number | null | Uint8Array)[] | undefined);
    if (results.length === 0) return [];

    const { columns, values } = results[0];
    return values.map((row) => {
      const obj: Record<string, unknown> = {};
      columns.forEach((col, i) => {
        obj[col] = row[i];
      });
      return obj;
    });
  }

  /** Delete a row by primary key value. */
  delete(key: unknown): void {
    const existing = this._getByPk(key);
    if (existing) {
      this._hash ^= this._hashRow(existing);
    }
    this._db.run(
      `DELETE FROM ${this.schema.name} WHERE ${this.pkCol} = ?`,
      [key] as (string | number | null | Uint8Array)[]
    );
    this._bump();
  }

  /** Delete all rows. Resets hash to 0. */
  clear(): void {
    this._db.run(`DELETE FROM ${this.schema.name}`);
    this._hash = 0;
    this._bump();
  }

  /** Current content hash (XOR of all row hashes). */
  getHash(): number {
    return this._hash;
  }

  /** Current version number. Incremented on every mutation. */
  getVersion(): number {
    return this._version;
  }

  /** Subscribe to mutations. Returns unsubscribe function. */
  subscribe(callback: Listener): () => void {
    this._listeners.add(callback);
    return () => {
      this._listeners.delete(callback);
    };
  }

  /**
   * React hook: subscribe to this table and re-run queryFn on mutations.
   * Uses useSyncExternalStore — only re-renders when this table changes.
   */
  useQuery<T>(queryFn: () => T): T {
    useSyncExternalStore(
      (cb) => this.subscribe(cb),
      () => this._version,
    );
    return queryFn();
  }

  /** Hash a row's values using djb2. */
  private _hashRow(row: Record<string, unknown>): number {
    const sorted = this._colNames.slice().sort();
    const str = JSON.stringify(sorted.map((c) => row[c]));
    return djb2(str);
  }

  /** Look up a single row by primary key. Returns null if not found. */
  private _getByPk(key: unknown): Record<string, unknown> | null {
    const results = this._db.exec(
      `SELECT * FROM ${this.schema.name} WHERE ${this.pkCol} = ?`,
      [key] as (string | number | null | Uint8Array)[]
    );
    if (results.length === 0 || results[0].values.length === 0) return null;
    const { columns, values } = results[0];
    const obj: Record<string, unknown> = {};
    columns.forEach((col, i) => {
      obj[col] = values[0][i];
    });
    return obj;
  }

  private _bump(): void {
    this._version++;
    for (const listener of this._listeners) {
      listener();
    }
  }
}

File: src/storage/table.py
"""Base Table abstraction for declarative SQLite table definitions."""

from __future__ import annotations

import json
import sqlite3
import threading
from dataclasses import dataclass, field


@dataclass
class Column:
    name: str
    type: str  # e.g. "INTEGER", "TEXT", "REAL"
    primary_key: bool = False
    autoincrement: bool = False
    not_null: bool = False
    default: str | None = None  # Raw SQL default, e.g. "datetime('now')"
    references: str | None = None  # e.g. "runs(id)"


@dataclass
class Index:
    name: str
    columns: list[str]


class Table:
    """Base class for a SQLite table.

    Subclasses declare `name`, `columns`, and `indexes`, then get
    automatic DDL generation and basic CRUD helpers.
    """

    name: str = ""
    columns: list[Column] = []
    indexes: list[Index] = []

    def __init__(self, conn: sqlite3.Connection, lock: threading.Lock):
        self._conn = conn
        self._lock = lock
        self._hash: int = 0
        self._broadcaster = None  # Optional: set via set_broadcaster()
        self._col_names: list[str] = [c.name for c in self.columns]
        pk_cols = [c for c in self.columns if c.primary_key]
        if len(pk_cols) != 1:
            raise ValueError(
                f'Table "{self.name}" must have exactly one primary key column, got {len(pk_cols)}'
            )
        self.pk_col: str = pk_cols[0].name

    def set_broadcaster(self, broadcaster):
        """Set a broadcaster to emit CDC ops on mutations."""
        self._broadcaster = broadcaster

    def _init_hash(self):
        """Compute hash from existing rows in the table."""
        with self._lock:
            self._rehash_unlocked()

    def create(self):
        """Generate and execute CREATE TABLE + CREATE INDEX statements."""
        col_defs = []
        for c in self.columns:
            parts = [c.name, c.type]
            if c.primary_key:
                parts.append("PRIMARY KEY")
            if c.autoincrement:
                parts.append("AUTOINCREMENT")
            if c.not_null:
                parts.append("NOT NULL")
            if c.default is not None:
                parts.append(f"DEFAULT ({c.default})")
            if c.references:
                parts.append(f"REFERENCES {c.references}")
            col_defs.append(" ".join(parts))

        ddl = f"CREATE TABLE IF NOT EXISTS {self.name} (\n"
        ddl += ",\n".join(f"    {d}" for d in col_defs)
        ddl += "\n);"

        idx_stmts = []
        for idx in self.indexes:
            cols = ", ".join(idx.columns)
            idx_stmts.append(
                f"CREATE INDEX IF NOT EXISTS {idx.name} ON {self.name}({cols});"
            )

        with self._lock:
            self._conn.execute(ddl)
            for stmt in idx_stmts:
                self._conn.execute(stmt)
            self._conn.commit()

    def insert(self, **kwargs) -> int:
        cols = list(kwargs.keys())
        placeholders = ", ".join("?" for _ in cols)
        col_names = ", ".join(cols)
        values = [kwargs[c] for c in cols]
        with self._lock:
            cursor = self._conn.execute(
                f"INSERT INTO {self.name} ({col_names}) VALUES ({placeholders})",
                values,
            )
            self._conn.commit()
            rowid = cursor.lastrowid
            # XOR in the new row's hash
            row = self._get_by_pk_unlocked(rowid)
            if row:
                self._hash ^= self._hash_row(row)
                self._emit_op("upsert", row=row)
            return rowid

    def insert_many(self, rows: list[dict]) -> list[int]:
        """Batch INSERT with one transaction, one read-back, one CDC emit."""
        if not rows:
            return []
        rowids: list[int] = []
        with self._lock:
            for row in rows:
                cols = list(row.keys())
                placeholders = ", ".join("?" for _ in cols)
                col_names = ", ".join(cols)
                values = [row[c] for c in cols]
                cursor = self._conn.execute(
                    f"INSERT INTO {self.name} ({col_names}) VALUES ({placeholders})",
                    values,
                )
                rowids.append(cursor.lastrowid)
            self._conn.commit()
            # Batch read-back and hash
            inserted_rows = []
            for rowid in rowids:
                full_row = self._get_by_pk_unlocked(rowid)
                if full_row:
                    self._hash ^= self._hash_row(full_row)
                    inserted_rows.append(full_row)
            # One batched CDC emit
            if inserted_rows:
                self._emit_ops("upsert", rows=inserted_rows)
        return rowids

    def select(
        self,
        where: str | None = None,
        params: list | None = None,
        order_by: str | None = None,
        columns: str = "*",
    ) -> list[dict]:
        query = f"SELECT {columns} FROM {self.name}"
        if where:
            query += f" WHERE {where}"
        if order_by:
            query += f" ORDER BY {order_by}"
        with self._lock:
            rows = self._conn.execute(query, params or []).fetchall()
            return [dict(r) for r in rows]

    def delete(self, where: str, params: list | None = None):
        with self._lock:
            # XOR out hashes of rows being deleted
            rows = self._conn.execute(
                f"SELECT * FROM {self.name} WHERE {where}", params or []
            ).fetchall()
            for row in rows:
                row_dict = dict(row)
                self._hash ^= self._hash_row(row_dict)
                self._emit_op("delete", key=row_dict[self.pk_col])
            self._conn.execute(f"DELETE FROM {self.name} WHERE {where}", params or [])
            self._conn.commit()

    def update(self, set_clause: str, where: str, params: list | None = None):
        with self._lock:
            all_params = params or []
            # Split params: count ? in set_clause for SET params, rest are WHERE params
            set_count = set_clause.count("?")
            where_params = all_params[set_count:]
            # Read full rows before update to XOR out old hashes
            old_rows = self._conn.execute(
                f"SELECT * FROM {self.name} WHERE {where}", where_params
            ).fetchall()
            for row in old_rows:
                self._hash ^= self._hash_row(dict(row))
            pk_vals = [dict(row)[self.pk_col] for row in old_rows]
            self._conn.execute(
                f"UPDATE {self.name} SET {set_clause} WHERE {where}", all_params
            )
            self._conn.commit()
            # Read back updated rows, XOR in new hashes, emit CDC
            updated_rows = []
            for pk in pk_vals:
                row = self._get_by_pk_unlocked(pk)
                if row:
                    self._hash ^= self._hash_row(row)
                    updated_rows.append(row)
            if updated_rows:
                self._emit_ops("upsert", rows=updated_rows)

    def upsert(self, **kwargs) -> int:
        """INSERT OR REPLACE a row. Requires explicit primary key value."""
        pk_val = kwargs.get(self.pk_col)
        if pk_val is None:
            raise ValueError(f'Row must include primary key column "{self.pk_col}"')
        with self._lock:
            # XOR out old row hash if it exists
            existing = self._get_by_pk_unlocked(pk_val)
            if existing:
                self._hash ^= self._hash_row(existing)
            cols = list(kwargs.keys())
            placeholders = ", ".join("?" for _ in cols)
            col_names = ", ".join(cols)
            values = [kwargs[c] for c in cols]
            self._conn.execute(
                f"INSERT OR REPLACE INTO {self.name} ({col_names}) VALUES ({placeholders})",
                values,
            )
            self._conn.commit()
            # XOR in the new row's hash (read back to get defaults)
            inserted = self._get_by_pk_unlocked(pk_val)
            if inserted:
                self._hash ^= self._hash_row(inserted)
                self._emit_op("upsert", row=inserted)
            return pk_val

    def upsert_many(self, rows: list[dict]):
        """Batch INSERT OR REPLACE. Each row must include the primary key."""
        if not rows:
            return
        for row in rows:
            pk_val = row.get(self.pk_col)
            if pk_val is None:
                raise ValueError(f'Row must include primary key column "{self.pk_col}"')
        with self._lock:
            upserted_rows = []
            for row in rows:
                pk_val = row[self.pk_col]
                existing = self._get_by_pk_unlocked(pk_val)
                if existing:
                    self._hash ^= self._hash_row(existing)
                cols = list(row.keys())
                placeholders = ", ".join("?" for _ in cols)
                col_names = ", ".join(cols)
                values = [row[c] for c in cols]
                self._conn.execute(
                    f"INSERT OR REPLACE INTO {self.name} ({col_names}) VALUES ({placeholders})",
                    values,
                )
                inserted = self._get_by_pk_unlocked(pk_val)
                if inserted:
                    self._hash ^= self._hash_row(inserted)
                    upserted_rows.append(inserted)
            self._conn.commit()
            if upserted_rows:
                self._emit_ops("upsert", rows=upserted_rows)

    def clear(self):
        """Delete all rows from the table."""
        with self._lock:
            self._conn.execute(f"DELETE FROM {self.name}")
            self._conn.commit()
            self._hash = 0
            self._emit_op("clear")

    def get_all_rows(self) -> list[dict]:
        """Return all rows as a list of dicts."""
        return self.select()

    def get_hash(self) -> int:
        """Current content hash (XOR of all row hashes)."""
        return self._hash

    def to_schema_dict(self) -> dict:
        """Serialize table schema to JSON-compatible dict matching frontend TableSchema."""
        columns = []
        for c in self.columns:
            col: dict = {"name": c.name, "type": c.type}
            if c.primary_key:
                col["primaryKey"] = True
            if c.autoincrement:
                col["autoincrement"] = True
            if c.not_null:
                col["notNull"] = True
            if c.default is not None:
                col["default"] = c.default
            columns.append(col)
        schema: dict = {"name": self.name, "columns": columns}
        if self.indexes:
            schema["indexes"] = [
                {"name": idx.name, "columns": idx.columns} for idx in self.indexes
            ]
        return schema

    # ── Private helpers ─────────────────────────────────────

    @staticmethod
    def _djb2(s: str) -> int:
        """djb2 hash over a string → unsigned 32-bit integer."""
        h = 5381
        for ch in s:
            h = ((h << 5) + h + ord(ch)) & 0xFFFFFFFF
        return h

    def _hash_row(self, row: dict) -> int:
        """Hash a row's values using djb2. Must match frontend implementation."""
        sorted_cols = sorted(self._col_names)
        s = json.dumps([row.get(c) for c in sorted_cols], separators=(",", ":"))
        return self._djb2(s)

    def _get_by_pk(self, key) -> dict | None:
        """Look up a single row by primary key. Acquires lock."""
        with self._lock:
            return self._get_by_pk_unlocked(key)

    def _get_by_pk_unlocked(self, key) -> dict | None:
        """Look up a single row by primary key. Caller must hold lock."""
        row = self._conn.execute(
            f"SELECT * FROM {self.name} WHERE {self.pk_col} = ?", [key]
        ).fetchone()
        return dict(row) if row else None

    def _emit_op(self, op: str, row: dict | None = None, key=None):
        """Emit a single CDC op to the broadcaster if set."""
        if self._broadcaster is not None:
            self._broadcaster.publish_op(self.name, op, row=row, key=key)

    def _emit_ops(self, op: str, rows: list[dict]):
        """Emit a batched CDC op to the broadcaster if set."""
        if self._broadcaster is not None:
            self._broadcaster.publish_ops(self.name, op, rows=rows)

    def _rehash_unlocked(self):
        """Recompute hash from all rows. Caller must hold lock."""
        h = 0
        rows = self._conn.execute(f"SELECT * FROM {self.name}").fetchall()
        for row in rows:
            h ^= self._hash_row(dict(row))
        self._hash = h
